src/hazelcore/Action_Exp.re

```diff
@@ -367,55 +302,101 @@ let zcase_of_scrut_and_suffix =
   };
 };
 
+let mk_text_operand_exp = (text: string): UHExp.operand => {
+  switch (TextShape.of_text(text)) {
+  | InvalidTextShape("") => EmptyHole(0)
+  | InvalidTextShape(s) => InvalidText(0, s)
```

Comment:
why are these metavars set to 0?

Reply:
I do the renumbering and hole fixing centrally. I moved thing around a little to clarify the process. I'm using a pattern I started using in the assistant branch; it's less relevant here but I think conceptually it makes sense. There, to avoid threading u_gen and ctx through a bunch of tiny functions I use to build syntax compositionally, I just fix/(re)number once after construction is complete.

Reply:
OK, but leave a comment on functions that generate mis-numbered holes so that other clients know that the caller is expected to renumber